### ISSUE TITLE
Counts for duplicate transcripts

### ISSUE THREAD

**Issue by jmacdon** [COMBINE-lab/salmon#301]
In the GRCh38 transcriptome you can get from Ensembl, there is at least one gene that is identical except for the fact that it appears both on the primary assembly and in an alternative locus. Ensembl gives this gene two ENSG IDs as well as two ENST IDs. Using R and an EnsemblDb, we have

> gns
             TXID          GENEID GENENAME           SEQNAME
1 ENST00000617214 ENSG00000273645  KBTBD11 CHR_HSCHR8_8_CTG1
2 ENST00000320248 ENSG00000176595  KBTBD11                 8

So two transcript and gene IDs, same gene name, but on chr8 and an alternate loci for chr8. If we read in the transcripts, the edit distance is zero:

> z
  A DNAStringSet instance of length 2
    width seq                                               names               
[1]  6710 AGCCCACCCGCCTGGCTGCGCGT...AAATCACGCAGAAAGTGCCAGTC ENST00000617214.2
[2]  6710 AGCCCACCCGCCTGGCTGCGCGT...AAATCACGCAGAAAGTGCCAGTC ENST00000320248.3
> stringDist(z)
                  ENST00000617214.2
ENST00000320248.3                 0

So identical lengths, identical sequences. But if we align using salmon (v0.10.2), using primarily defaults of

salmon quant -i <index> -l A -1 <fastq1> -2 <fastq2> -o <output>

we get 

$ grep ENST00000617214 siR_*/quant.sf  | cut -f 1,5
siR_01_quant/quant.sf:ENST00000617214	326.951111
siR_02_quant/quant.sf:ENST00000617214	500.506648
siR_03_quant/quant.sf:ENST00000617214	430.630303
siR_04_quant/quant.sf:ENST00000617214	466.048848
siR_NC01_quant/quant.sf:ENST00000617214	923.891639
siR_NC02_quant/quant.sf:ENST00000617214	1290.138889
siR_NC03_quant/quant.sf:ENST00000617214	942.277397
siR_NC04_quant/quant.sf:ENST00000617214	904.379677

$ grep ENST00000320248 siR_*/quant.sf  | cut -f 1,5
siR_01_quant/quant.sf:ENST00000320248	1218.048889
siR_02_quant/quant.sf:ENST00000320248	1795.493352
siR_03_quant/quant.sf:ENST00000320248	1579.369697
siR_04_quant/quant.sf:ENST00000320248	1587.951152
siR_NC01_quant/quant.sf:ENST00000320248	945.108361
siR_NC02_quant/quant.sf:ENST00000320248	549.861111
siR_NC03_quant/quant.sf:ENST00000320248	1001.722603
siR_NC04_quant/quant.sf:ENST00000320248	895.620323

Where the counts within group (first four samples vs last four samples) are relatively consistent, but the direction 'flips' for the two transcripts, being higher in the second group for the first transcript and vice versa for the second.

The counts for each library are pretty similar, being about 25M or so:

> counts <- tximport(samps$files, "salmon", tx2gene = tx2gene)
reading in files with read_tsv
1 2 3 4 5 6 7 8 
transcripts missing from tx2gene: 1600
summarizing abundance
summarizing counts
summarizing length
> colSums(counts$counts)/1e6
[1] 19.81646 25.65678 26.80554 23.44890 25.83873 24.82431 25.84806 24.10782

Anyway, long story short, it seems weird that we would be getting such consistent (within group) results that end up being so discordant between the two (identical) transcripts to which we have aligned. 

Any suggestions? Or is the take-home that we shouldn't use Ensembl transcripts because they have these duplicates?

**Comment by rob-p** [comment]
Hi James,

  Could you provide the input transcriptome file you're using for quantification here?  This is somewhat surprising, as the indexer should, by default, remove duplicate transcripts from the index.  So, the fact that we have 2 (sequence) duplicate transcripts being quantified here is the first question to address.  I'll be happy to take a look into it, but I want to start with exactly the same input set you're working with.

**Comment by jmacdon** [comment]
Hi Rob,

A colleague did the alignment, using some really old Ensembl version (ftp://ftp.ensembl.org/pub/release-79/fasta/homo_sapiens/cdna/Homo_sapiens.GRCh38.cdna.all.fa.gz). For the example I gave I used the current version, which should be +/- the same (ftp://ftp.ensembl.org/pub/release-94/fasta/homo_sapiens/cdna/Homo_sapiens.GRCh38.cdna.all.fa.gz).

Thanks!

Jim

**Comment by rob-p** [comment]
Hi Jim,

  So, this is strange.  I just indexed the GRCh38 cdna file, [here](https://github.com/COMBINE-lab/salmon/files/2474696/duplicate_clusters.tsv.zip) is the `duplicate_clusters.tsv` file from the index.  Indeed, it contains exactly the transcript you point out.

```
rob@computer:~/salmon/build$ grep "ENST00000617214.2" james_macdonald/index/duplicate_clusters.tsv
ENST00000320248.3       ENST00000617214.2
```
telling us that the transcripts are duplicates, `ENST00000320248.3` was retained and `ENST00000617214.2` was not indexed (and so should not appear in the `quant.sf` file).  Did you, by any chance, use an older salmon _index_?  Could you try re-indexing with the version you have?

Best,
Rob

**Comment by jmacdon** [comment]
Hi Rob,

Those transcripts aren't in the duplicate_clusters file. Is there a way to determine what version of salmon was used to generate the index?

**Comment by rob-p** [comment]
Hi James,

  Surprisingly, it looks like I don't actually log exactly the version of salmon used in the index.  In the index directory, the `header.json` will give you some info.  It will tell you the *index* version, but many versions of salmon can correspond to the same *index* version.  The info in that file might help to discriminate a little bit.  Can you try *indexing* with v0.10.2 (the one you are aligning with)?

**Comment by wedelma** [comment]
Hi Rob,

Thanks to both you and Jim for helping with this, since I'm the one who actually need some help. I've tried indexed using v0.11.0 then the latest version. Both, with the same transcriptome you tried doesn't result in duplicates when I grep "ENST00000617214.2". Telling me I'm getting a different result from your grep. I don't know why. I will be downloading the latest version of the human transcriptome and using latest salmon to index, and I should see these duplicate transcripts in the duplicate clusters file, like you did, likely resolving my issue. We had older data which is why I used an older version of the human transcriptome, which I will not be doing anymore. I'll keep you posted and thanks for the help!

**Comment by jmacdon** [comment]
Hi Bill,

You probably shouldn't be including the transcript version number (the
trailing . 2) when using grep, to be safe. It's possible that you could
have a different transcript version so there isn't any profit in being so
exact.

Jim

**Comment by wedelma** [comment]
Right I've been checking it without the trailing character as well, I changed to leaving that out. 

Oddly, my index doesn't have the decimals at all, just continuous ENSTxx, unlike the examples you and Rob provided. I redownloaded the new release, same result. No duplicate in my grep query, which would lead to getting the quant for both transcripts. Not being filtered by salmon as duplicates. I also  downloaded the same version in Rob's example above, indexed with latest salmon,  with the same result. At a loss right now. I don't know why this is.

**Comment by rob-p** [comment]
Hi Bill,

  Strange indeed! Can you share the log salmon made during indexing and the `duplicate_clusters.tsv` file you do get?

Thanks,
Rob

**Comment by wedelma** [comment]
Here is the duplicate clusters (extension edited for uploading as .txt) and log. 
[quasi_index.log](https://github.com/COMBINE-lab/salmon/files/2474978/quasi_index.log)


[duplicate_clusters (copy).txt](https://github.com/COMBINE-lab/salmon/files/2474977/duplicate_clusters.copy.txt)

**Comment by rob-p** [comment]
Ok, in fact, when I check with the "older" reference (from ftp://ftp.ensembl.org/pub/release-79/fasta/homo_sapiens/cdna/Homo_sapiens.GRCh38.cdna.all.fa.gz), I get that these transcripts are _not_ duplicates of each other.  To verify it's not salmon's problem, I also checked with [seqkit's rmdup command](https://bioinf.shenwei.me/seqkit/usage/#rmdup), and, in fact, in the older cdna file, these transcripts appear to have distinct sequence.

**Comment by jmacdon** [comment]
Well that explains things. It's still interesting that the apparent abundance flips sign - at some point EBI/EMBL and NCBI for that matter decided the sequences are identical - but using the old version of the transcriptome we do get an apparent difference that may point to differential transcript abundance?

Anyway, thanks for all your help, Rob.

**Comment by rob-p** [comment]
Hi James,

  Indeed!  No problem.  Looking at the older annotation, it seems that the transcripts have slightly different lengths:

```
>ENST00000617214 ensembl:known chromosome:GRCh38:CHR_HSCHR8_8_CTG1:1973878:2006943:1 gene:ENSG00000273645 gene_biotype:protein_coding transcript_biotype:protein_coding

>ENST00000320248 ensembl_havana_transcript:known chromosome:GRCh38:8:1973878:2006936:1 gene:ENSG00000176595 gene_biotype:protein_coding transcript_biotype:protein_coding
```

I don't know if that's the _only_ difference between them or if they are spliced differently in the middle etc.  Both salmon and seqkit only check for *exact* sequence duplication.  If you wanted to try to dig more into what might be causing the differences you might consider `<salmon_quant_folder>/aux_info/ambig_info.tsv` which lists the transcripts in the same order as `quant.sf` but gives for each the number of uniquely mapping reads and the total number of potential reads.  You could see if the unique evidence for the txps changes between conditions.  However, given that Ensembl has changed their mind about these guys, maybe that's not a rabbit hole worth going down.  Anyway, glad I could help!
